File: apps/android/app/src/main/java/com/exponential/app/ui/components/TreeGuidesOverlay.kt
package com.exponential.app.ui.components

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.exponential.app.domain.TreeGuide
import com.exponential.app.domain.TreeGuides
import com.exponential.app.ui.theme.GlassTokens

// EXP-965: ONE drawing of the nested-list connector, applied by every list
// that nests (the Agent page's Running band and its Recent sheet, the stack
// overlay's Runs and Pull requests, the Reviews stack members). The SHAPE is
// pure (`domain/TreeGuides.kt`, mirrored ×4); this file only paints it.

/** The connector's stroke — the shared hairline, never a literal colour. */
private val GuideWidth = 1.dp

/** The elbow's turn radius — a tight one (EXP-998: 5 read as a bulge on a
 *  28dp row), the ×4 number. */
private val GuideRadius = 3.dp

/**
 * Draw [guide] in the gutter bands to the LEFT of this row's content. Apply it
 * BEFORE the indent padding, so the draw area still spans the gutters the
 * lines live in ([TreeGuidesRow] does exactly that).
 *
 * [gap] is the LIST's row spacing: every vertical that starts at the row's top
 * edge starts that much ABOVE it (nothing clips the row, so the ink lands in
 * the list's own gap), while a pass-through or a tee still ends at the bottom
 * edge — the next row's extension is what covers the gap. Without it a spaced
 * list drew the branch as a dashed ladder. The ×4 rule; the pure shape
 * (`domain/TreeGuides.kt`) knows nothing about it.
 */
fun Modifier.treeGuides(guide: TreeGuide?, gap: Dp = 0.dp): Modifier {
    if (guide == null || guide.isEmpty) return this
    return drawBehind {
        val indent = TreeGuides.INDENT_DP.dp.toPx()
        val stroke = Stroke(width = GuideWidth.toPx())
        val radius = GuideRadius.toPx()
        val midY = size.height / 2f
        // The row above ends `gap` up there: start every downward line from it.
        val top = -gap.toPx()
        fun centreOf(level: Int) = indent * level + indent / 2f

        // An ancestor whose subtree carries on: a straight full-height line.
        guide.passThrough.forEach { level ->
            val path = Path().apply {
                moveTo(centreOf(level), top)
                lineTo(centreOf(level), size.height)
            }
            drawPath(path, GlassTokens.StrokeStrong, style = stroke)
        }

        val elbow = guide.elbowAt ?: return@drawBehind
        val x = centreOf(elbow)
        val right = indent * (elbow + 1)
        val path = Path().apply {
            if (guide.tee) {
                // A later sibling follows: the vertical runs the whole height,
                // unbroken, and the corner below branches off it (EXP-998:
                // the same rounded branch as the last child, ×4).
                moveTo(x, top)
                lineTo(x, size.height)
                moveTo(x, midY - radius)
            } else {
                moveTo(x, top)
                lineTo(x, midY - radius)
            }
            arcTo(
                rect = Rect(x, midY - 2f * radius, x + 2f * radius, midY),
                startAngleDegrees = 180f,
                sweepAngleDegrees = -90f,
                forceMoveTo = false,
            )
            lineTo(right, midY)
        }
        drawPath(path, GlassTokens.StrokeStrong, style = stroke)
    }
}

/**
 * One nested row: [TreeGuides.INDENT_DP] of indent per level with [guide]
 * drawn in the gutters it leaves behind. Every nesting list wraps its row in
 * this instead of a bare indent padding. [gap] = the list's row spacing, so
 * the branch bridges it.
 */
@Composable
fun TreeGuidesRow(
    depth: Int,
    guide: TreeGuide?,
    modifier: Modifier = Modifier,
    gap: Dp = 0.dp,
    content: @Composable () -> Unit,
) {
    Box(
        modifier = modifier
            .fillMaxWidth()
            .treeGuides(guide, gap)
            .padding(start = (TreeGuides.INDENT_DP * depth).dp),
    ) {
        content()
    }
}

File: apps/ios/ExpUI/Sources/TreeGuidesOverlay.swift
import ExpCore
import SwiftUI

/// EXP-965: the tree CONNECTOR behind a nested row — the one drawing of
/// `TreeGuides` (ExpCore owns the rule, this owns the ink). A 1pt
/// `strokeStrong` line down the parent level's gutter, a rounded elbow into
/// the row's leading glyph, a tee where a sibling follows, and a straight
/// line for every ancestor level whose subtree carries on below.
///
/// Sized by whatever it backs: the gutters are measured from the row's
/// leading edge (`base` = the row's own content inset, so the vertical lands
/// on the PARENT row's leading glyph centre) and the elbow turns at the row's
/// vertical centre.
///
/// `gap` is the LIST's row spacing, the ×4 rule: every vertical that starts
/// at the row's top edge starts `gap` ABOVE it instead (drawn outside the row
/// bounds), while a tee or a pass-through still ends at the BOTTOM edge — the
/// next row's extension is what covers the gap. Without it a spaced list
/// draws a dashed branch, one break per row.
public struct TreeGuidesOverlay: View {
    private let guide: TreeGuide
    private let indent: CGFloat
    private let base: CGFloat
    private let gap: CGFloat

    /// The default `base`: every flat list row pads its content 12pt, and the
    /// fold chevron beside it is 14 wide — so a parent's glyph centre sits
    /// `12 + 7` in, exactly the first gutter's centre.
    public static let rowContentInset: CGFloat = 12

    public init(
        guide: TreeGuide,
        indent: CGFloat = TreeGuides.indentPerLevel,
        base: CGFloat = TreeGuidesOverlay.rowContentInset,
        gap: CGFloat = 0
    ) {
        self.guide = guide
        self.indent = indent
        self.base = base
        self.gap = gap
    }

    public var body: some View {
        // The canvas grows UP by the list's row spacing and is drawn outside
        // the row it backs, so the branch runs through the gap above it.
        GeometryReader { proxy in
            Canvas { context, size in
                draw(in: context, size: size)
            }
            .frame(height: proxy.size.height + gap)
            .offset(y: -gap)
        }
        .allowsHitTesting(false)
        .accessibilityHidden(true)
    }

    /// `size` spans the gap ABOVE the row plus the row itself: the row's top
    /// edge sits at `gap`, its bottom at `size.height`.
    private func draw(in context: GraphicsContext, size: CGSize) {
        guard !guide.isEmpty else { return }
        let middle = gap + (size.height - gap) / 2
        var path = Path()
        for level in guide.passThrough {
            let x = centre(of: level)
            path.move(to: CGPoint(x: x, y: 0))
            path.addLine(to: CGPoint(x: x, y: size.height))
        }
        if let level = guide.elbowAt {
            let x = centre(of: level)
            let radius = min(TreeGuides.elbowRadius, middle - gap)
            path.move(to: CGPoint(x: x, y: 0))
            path.addLine(to: CGPoint(x: x, y: middle - radius))
            path.addQuadCurve(
                to: CGPoint(x: x + radius, y: middle),
                control: CGPoint(x: x, y: middle)
            )
            path.addLine(to: CGPoint(x: base + indent * CGFloat(level + 1), y: middle))
            // Not the last child: the vertical carries on to the next one.
            if guide.tee {
                path.move(to: CGPoint(x: x, y: 0))
                path.addLine(to: CGPoint(x: x, y: size.height))
            }
        }
        context.stroke(path, with: .color(GlassTokens.strokeStrong), lineWidth: 1)
    }

    /// The gutter band for `level` runs `base + indent * level` →
    /// `+ indent`; its centre is where the parent's glyph sits.
    private func centre(of level: Int) -> CGFloat {
        base + indent * CGFloat(level) + indent / 2
    }
}

extension View {
    /// EXP-965: indent this row to its depth and draw its connector — the ONE
    /// way a nested list row is offset (every site used to type its own 14).
    /// `gap` is the list's own row spacing, which the branch runs through.
    public func treeGuides(
        _ guide: TreeGuide,
        base: CGFloat = TreeGuidesOverlay.rowContentInset,
        gap: CGFloat = 0
    ) -> some View {
        padding(.leading, CGFloat(guide.depth) * TreeGuides.indentPerLevel)
            .background(TreeGuidesOverlay(guide: guide, base: base, gap: gap))
    }
}
